velox/substrait/VeloxToSubstraitType.cpp

```diff
@@ -0,0 +1,336 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/substrait/VeloxToSubstraitType.h"
+
+#include "velox/expression/Expr.h"
+
+namespace facebook::velox::substrait {
+
+::substrait::Type VeloxToSubstraitTypeConvertor::toSubstraitType(
+    const velox::TypePtr& vType,
+    ::substrait::Type* sType) {
+  switch (vType->kind()) {
+    case velox::TypeKind::BOOLEAN: {
+      sType->set_allocated_bool_(new ::substrait::Type_Boolean());
+      return *sType;
+    }
+    case velox::TypeKind::TINYINT: {
+      sType->set_allocated_i8(new ::substrait::Type_I8());
+      return *sType;
+    }
+    case velox::TypeKind::SMALLINT: {
+      sType->set_allocated_i16(new ::substrait::Type_I16());
+      return *sType;
+    }
+    case velox::TypeKind::INTEGER: {
+      sType->set_allocated_i32(new ::substrait::Type_I32());
+      return *sType;
+    }
+    case velox::TypeKind::BIGINT: {
+      sType->set_allocated_i64(new ::substrait::Type_I64());
+      return *sType;
+    }
+    case velox::TypeKind::REAL: {
+      sType->set_allocated_fp32(new ::substrait::Type_FP32());
+      return *sType;
+    }
+    case velox::TypeKind::DOUBLE: {
+      sType->set_allocated_fp64(new ::substrait::Type_FP64());
+      return *sType;
+    }
+    case velox::TypeKind::VARCHAR: {
+      sType->set_allocated_varchar(new ::substrait::Type_VarChar());
+      return *sType;
+    }
+    case velox::TypeKind::VARBINARY: {
+      sType->set_allocated_binary(new ::substrait::Type_Binary());
+      return *sType;
+    }
+    case velox::TypeKind::TIMESTAMP: {
+      sType->set_allocated_timestamp(new ::substrait::Type_Timestamp());
+      return *sType;
+    }
+    case velox::TypeKind::ARRAY: {
+      ::substrait::Type_List* sTList = new ::substrait::Type_List();
```

Comment:
What's the motivation for using raw pointers and not smart pointers? If there is an exception in say toSubstraitType on L72, the memory allocated here will be leaked.

Reply:
can change the parameter in L72(toSubstraitType ) to shared_ptr.  But the parameter in  `sType->set_allocated_list(sTList)` need to be raw pointer. We cannot do conversion from 'std::shared_ptr< ::substrait::Type_List>' to '::substrait::Type_List *' if we use smart pointer for sTList.

Reply:
Hmm... how do you suggest to fix the memory leak in the face of exceptions then? CC: @rui-mo

Reply:
does this make sense ? 
`      auto sTList = std::make_shared<::substrait::Type_List>();
      const TypePtr vElementType = vType->asArray().elementType();

      toSubstraitType(
          vElementType,
          std::make_shared<::substrait::Type>(*(sTList->mutable_type())));
      
      sType->set_allocated_list(sTList.get());`

Reply:
does this make sense?
     
     auto sTList = std::make_shared<::substrait::Type_List>();
      const TypePtr vElementType = vType->asArray().elementType();

      toSubstraitType(
          vElementType,
          std::make_shared<::substrait::Type>(*(sTList->mutable_type())));
      
      sType->set_allocated_list(sTList.get());

Reply:
I don't have experience working with Protobuf. A quick read suggests that the common pattern is to use arena-based allocations: https://developers.google.com/protocol-buffers/docs/reference/arenas#messageclass

Therefore, I assume we need to discuss whether we'll use that or not.

The example above allocates Type_List and stores it in a std::shared_ptr. This ensures that Type_List will get deallocated automatically when reference count in the shared_ptr goes to zero. 

I assume sTList->mutable_type() allocates new memory within the Type_List and I further assume that this memory will be freed automatically when Type_List is destroyed. If that's the case, there is no need to wrap sTList->mutable_type() in a shared_ptr. Instead we can pass a reference (not a raw pointer) to toSubstraitType.

> sType->set_allocated_list(sTList.get());

I'm not sure this works. Here we pass a raw pointer to sType. The memory pointed by that will be freed when sTList goes out of scope leaving sType referencing deallocated memory. I expect this will cause a crash.

Reply:
The raw pointer can be got from a shared_ptr with something like "&(*ptr)", but this would break the reference count logic of this shared_ptr and can cause crash.
I think use arena-based allocations is a good suggestion and the shared_ptr is not needed then.

Reply:
fix it